internal/proxyservice/sshagent/proxy_service.go

```diff
@@ -0,0 +1,162 @@
+package sshagent
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+
+	validation "github.com/go-ozzo/ozzo-validation"
+	"golang.org/x/crypto/ssh/agent"
+
+	"github.com/cyberark/secretless-broker/internal"
+	"github.com/cyberark/secretless-broker/pkg/secretless/log"
+)
+
+type proxyService struct {
+	done                bool
+	listener            net.Listener
+	logger              log.Logger
+	keyring             agent.Agent
+	retrieveCredentials internal.CredentialsRetriever
+}
+
+// NewProxyService constructs a new instance of a SSH Agent ProxyService. The
+// constructor takes a Listener, Logger and CredentialResolver.
+// An SSH Agent ProxyService is able to server SSH Agent requests with Credentials
+func NewProxyService(
+	listener net.Listener,
+	logger log.Logger,
+	retrieveCredentials internal.CredentialsRetriever,
+) (internal.Service, error) {
+	errors := validation.Errors{}
+
+	if retrieveCredentials == nil {
+		errors["retrieveCredentials"] = fmt.Errorf("retrieveCredentials cannot be nil")
+	}
+	if logger == nil {
+		errors["logger"] = fmt.Errorf("logger cannot be nil")
+	}
+	if listener == nil {
+		errors["listener"] = fmt.Errorf("listener cannot be nil")
+	}
+
+	if err := errors.Filter(); err != nil {
+		return nil, err
+	}
+
+	return &proxyService{
+		retrieveCredentials: retrieveCredentials,
+		listener:            listener,
+		logger:              logger,
+		done:                false,
+		keyring:             agent.NewKeyring(),
+	}, nil
+}
+
+func parseKey(pemStr []byte) (interface{}, error) {
+	block, _ := pem.Decode(pemStr)
+	if block == nil {
+		return nil, fmt.Errorf("failed to decode PEM block")
+	}
+
+	switch block.Type {
+	case "RSA PRIVATE KEY":
+		return x509.ParsePKCS1PrivateKey(block.Bytes)
+	case "EC PRIVATE KEY":
+		return x509.ParseECPrivateKey(block.Bytes)
+	default:
+		return nil, fmt.Errorf("unsupported key type %q", block.Type)
+	}
+}
+
+// Start initiates the net.Listener to listen for incoming connections
+func (proxy *proxyService) Start() error {
+	logger := proxy.logger
+
+	logger.Infof("Starting service")
+
+	if proxy.done {
+		return fmt.Errorf("cannot call Start on stopped ProxyService")
+	}
+
+	// TODO: this proxy service only fetches credentials at the outset
+	//   so can not cope with rotation
+	backendCredentials, err := proxy.retrieveCredentials()
+	defer internal.ZeroizeCredentials(backendCredentials)
+	if err != nil {
+		return fmt.Errorf("failed on retrieve credentials: %s", err)
+	}
+
+	key := agent.AddedKey{}
+
+	if rsa, ok := backendCredentials["rsa"]; ok {
+		key.PrivateKey, err = parseKey(rsa)
+		if err != nil {
+			return err
+		}
+	}
+	if ecdsa, ok := backendCredentials["ecdsa"]; ok {
+		key.PrivateKey, err = parseKey(ecdsa)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
is "ecdsa" supposed to override "rsa" or is it never the case that both can exist?  if the latter, should use switch statement?  (if.. else is an alternative but switch is clearer imo and switch over if... else should even be a style guide rule).

Reply:
it should be one or the other. changed